scale/scale_test.go

```diff
@@ -0,0 +1,113 @@
+package scale
+
+import (
+	"github.com/stretchr/testify/assert"
+	"testing"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type MockDeployment struct {
+	client *MockKubeClient
+}
+
+type MockKubeClient struct {
+	// stores the state of Deployment as if the api server did
+	Deployment *extensions.Deployment
+}
+
+func (m *MockDeployment) Get(name string) (*extensions.Deployment, error) {
+	return m.client.Deployment, nil
+}
+
+func (m *MockDeployment) Update(deployment *extensions.Deployment) (*extensions.Deployment, error) {
+	m.client.Deployment.Spec.Replicas = deployment.Spec.Replicas
+	return m.client.Deployment, nil
+}
+
+func (m *MockDeployment) List(opts api.ListOptions) (*extensions.DeploymentList, error) {
+	return nil, nil
+}
+
+func (m *MockDeployment) Delete(name string, options *api.DeleteOptions) error {
+	return nil
+}
+
+func (m *MockDeployment) Create(*extensions.Deployment) (*extensions.Deployment, error) {
+	return nil, nil
+}
+
+func (m *MockDeployment) UpdateStatus(*extensions.Deployment) (*extensions.Deployment, error) {
+	return nil, nil
+}
+
+func (m *MockDeployment) Watch(opts api.ListOptions) (watch.Interface, error) {
+	return nil, nil
+}
+
+func (m *MockDeployment) Rollback(*extensions.DeploymentRollback) error {
+	return nil
+}
+
+func (m *MockKubeClient) Deployments(namespace string) kclient.DeploymentInterface {
+	return &MockDeployment{
+		client: m,
+	}
+}
+
+func NewMockKubeClient() *MockKubeClient {
+	return &MockKubeClient{
+		Deployment: &extensions.Deployment{
+			Spec: extensions.DeploymentSpec{
+				Replicas: 3,
+			},
+		},
+	}
+}
+
+func NewMockPodAutoScaler(kubernetesDeploymentName string, kubernetesNamespace string, max int, min int) *PodAutoScaler {
+	mockClient := NewMockKubeClient()
+
+	return &PodAutoScaler{
+		Client:     mockClient,
+		Min:        min,
+		Max:        max,
+		Deployment: kubernetesDeploymentName,
+		Namespace:  kubernetesNamespace,
+	}
+}
+
+func TestScaleUp(t *testing.T) {
+	p := NewMockPodAutoScaler("test", "test", 5, 1)
+
+	_ = p.ScaleUp()
+	deployment, _ := p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(4), deployment.Spec.Replicas)
+	_ = p.ScaleUp()
+	deployment, _ = p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(5), deployment.Spec.Replicas)
+
+	err := p.ScaleUp()
+	assert.NotNil(t, err)
+	deployment, _ = p.Client.Deployments("test").Get("test")
+	assert.Equal(t, int32(5), deployment.Spec.Replicas)
```

Comment:
after a pause and contemplation, i can see why this is still 5 instead of 6, but here's where a "why" comment would make reading the code much faster. that comment should probably be a few lines up to explain this block as a whole